feat: Allow to expand until certain condition is met

ghostbuster91 opened the pull request:
This is my attempt to solve #2789 

With this change I can now define the following mapping:
```lua
local function stop_expansion(_, node)
    return false, stop_expansion
end
local function expand_until_non_single(count, node)
    local cwd = node.link_to or node.absolute_path
    local handle = vim.loop.fs_scandir(cwd)
    if not handle then
        return false, stop_expansion
    end
    local status = git.load_project_status(cwd)
    populate_children(handle, cwd, node, status)
    local child_folder_only = explorer_node.has_one_child_folder(node) and node.nodes[1]
    if count > 1 and not child_folder_only then
        return true, stop_expansion
    elseif child_folder_only then
        return true, expand_until_non_single
    else
        return false, stop_expansion
    end
end
map('n', 'E', function()
    api.tree.expand_all(nil, expand_until_non_single)
end, opts("Expand until not single"))
```

This works as expected however it has some downsides:
1. the api is very flexible but at the same time quite complex. I had to do it in this way because the `expand_all` function will not expand the last directory. It stops as soon as `should_expand` returns `false` but I wanted the last directory to be expanded.
2. ~I had to copy `populate_children` method which is not a part of public api~ replaced with `explorer.explore(node, status)`

Btw if this got accepted in some form I am going to replace the default expand behavior in my config with this, since I find it so useful.

alex-courtis commented:
I love the concept and it is something we should do. Tried your example after a lot of hacking and it looks to work nicely.

- [ ] Does nothing when called from the root node
- [ ] Needs documentation

Use of private functions in your example aside, this is far too complex and not approachable for users. Having to populate the children is the key problem.

Suggestion: come up with a simple use case and example - one that would go in the documentation. We can start simplifying from there.

Perhaps I'm misunderstanding: why does this need to be an iterator? I was imagining the user would supply a simple boolean returning function to test against a node.

alex-courtis commented:
Any updates on this one @ghostbuster91 ?

ghostbuster91 commented:
Sorry, I was quite busy. I am going to take a look at it. 

> I love the concept and it is something we should do.

I am glad to hear that!

> Perhaps I'm misunderstanding: why does this need to be an iterator? I was imagining the user would supply a simple boolean returning function to test against a node.

Because the simple `fn: node => boolean` is insufficient in my case. Take a look at the following example:

```
a
└── b
    └── c
        ├── c2
        │   └── d2 (file)
        └── d (file)
```

When expanding `a` I would like to expand `b` and `c`. 
However `c` has multiple children and also directories inside so it wouldn't pass a simple `should_expand` function. 

In my case the condition is following:
Keep expanding until you reach a node that has more than one children but expand this node as well. In other words the expansion is inclusive. 

I suppose this could be expressed in simpler way for example by using a flag - `inclusive=true` on the user side. 

> come up with a simple use case and example - one that would go in the documentation. 

Sure, let's conclude first what to do with the optionally inclusive expansion.

ghostbuster91 commented:
In the end I was able to simply replace the iterator with accessing node's siblings via `node.parent.nodes`. With this the user side function for my particular use case looks as follow:
```lua
local function expand_until_non_single(count, node, populate_node)
    populate_node(node)
    if node.nodes == nil or not node.parent.open then
        return false
    end
    local has_one_child_folder = explorer_node.has_one_child_folder(node)
    local has_no_siblings = explorer_node.has_one_child_folder(node.parent)
    if has_one_child_folder and count == 0 then
        return true
    elseif has_no_siblings then
        return true
    else
        return false
    end
end
```

The CI keeps failing with following error:
```
{
    "file:///home/runner/work/nvim-tree.lua/nvim-tree.lua/lua/nvim-tree/actions/tree/modifiers/expand-all.lua": [
        {
            "code": "redundant-parameter",
            "message": "This function expects a maximum of 2 argument(s) but instead it is receiving 3.",
            "range": {
                "end": {
                    "character": 43,
                    "line": 33
                },
                "start": {
                    "character": 37,
                    "line": 33
                }
            },
            "severity": 2,
            "source": "Lua Diagnostics."
        }
    ]
}
```
Any idea how to fix it?

ghostbuster91 commented on lua/nvim-tree/explorer/node.lua:
```diff
@@ -36,7 +36,7 @@ end
 ---@return boolean
 function M.has_one_child_folder(node)
-  return #node.nodes == 1 and node.nodes[1].nodes and vim.loop.fs_access(node.nodes[1].absolute_path, "R") or false
+  return node.nodes ~= nil and #node.nodes == 1 and node.nodes[1].nodes and vim.loop.fs_access(node.nodes[1].absolute_path, "R") or false
```
**thought**: guarding against calling this function with a file node

alex-courtis commented:
This minimal setup worked nicely:

```lua
vim.g.loaded_netrw = 1
vim.g.loaded_netrwPlugin = 1

vim.cmd([[set runtimepath=$VIMRUNTIME]])
vim.cmd([[set packpath=/tmp/nvt-min/site]])
local package_root = "/tmp/nvt-min/site/pack"
local install_path = package_root .. "/packer/start/packer.nvim"
local function load_plugins()
  require("packer").startup({
    {
      "wbthomason/packer.nvim",
      "nvim-tree/nvim-tree.lua",
      "nvim-tree/nvim-web-devicons",
      -- ADD PLUGINS THAT ARE _NECESSARY_ FOR REPRODUCING THE ISSUE
    },
    config = {
      package_root = package_root,
      compile_path = install_path .. "/plugin/packer_compiled.lua",
      display = { non_interactive = true },
    },
  })
end
if vim.fn.isdirectory(install_path) == 0 then
  print("Installing nvim-tree and dependencies.")
  vim.fn.system({ "git", "clone", "--depth=1", "https://github.com/wbthomason/packer.nvim", install_path })
end
load_plugins()
require("packer").sync()
vim.cmd([[autocmd User PackerComplete ++once echo "Ready!" | lua setup()]])
vim.opt.termguicolors = true
vim.opt.cursorline = true

local api = require("nvim-tree.api")
local lib = require("nvim-tree.lib")
local explorer_node = require("nvim-tree.explorer.node")

local function expand_until_non_single(count, node, populate_node)
  populate_node(node)
  if node.nodes == nil or not node.parent.open then
    return false
  end
  local has_one_child_folder = explorer_node.has_one_child_folder(node)
  local has_no_siblings = explorer_node.has_one_child_folder(node.parent)
  if has_one_child_folder and count == 0 then
    return true
  elseif has_no_siblings then
    return true
  else
    return false
  end
end

local function expand_until()
  local node = api.tree.get_node_under_cursor()
  if node.open then
    lib.expand_or_collapse(node, nil)
  else
    if node.nodes then
      api.tree.expand_all(node, expand_until_non_single)
    else
      api.node.open.edit(node)
    end
  end
end

local function on_attach(bufnr)
  local function opts(desc)
    return { desc = "nvim-tree: " .. desc, buffer = bufnr, noremap = true, silent = true, nowait = true }
  end

  api.config.mappings.default_on_attach(bufnr)

  vim.keymap.set("n", "O", expand_until, opts("Expand until"))
end

_G.setup = function()
  require("nvim-tree").setup({
    on_attach = on_attach,
  })
end
```

alex-courtis commented:
Suggestions:

```lua
lib.expand_or_collapse(node, nil)
```

Replace with `node.open.edit` after testing for children i.e. `node.nodes`

```lua
  local has_one_child_folder = explorer_node.has_one_child_folder(node)
  local has_no_siblings = explorer_node.has_one_child_folder(node.parent)
```

Write your own test in this case. It will be a lot simpler as you have tested most of the conditions already.

alex-courtis commented:
> Any idea how to fix it?

![20240902_121129](https://github.com/user-attachments/assets/ebecf0f8-99bc-4610-ad42-20173fb6ce2b)

status is not a parameter for `function Explorer:expand(node)`, looks like it's unused.

alex-courtis commented:
![20240902_121527](https://github.com/user-attachments/assets/01251e4c-5421-442e-ae05-a3654f5bb234)

That parameter is not used and thus can be removed.

alex-courtis commented on lua/nvim-tree/actions/tree/modifiers/expand-all.lua:
```diff
@@ -64,9 +79,10 @@ local function gen_iterator()
 
 ---@param base_node table
-function M.fn(base_node)
+function M.fn(base_node, expand_until)
```
Please use an options table similar to `api.tree.resize`:
* add a class to `api.lua` to describe this
* docuement this at `*nvim-tree-api.tree.expand_all()*`

alex-courtis commented on lua/nvim-tree/actions/tree/modifiers/expand-all.lua:
```diff
@@ -3,6 +3,7 @@ local renderer = require "nvim-tree.renderer"
 local Iterator = require "nvim-tree.iterators.node-iterator"
 local notify = require "nvim-tree.notify"
 local lib = require "nvim-tree.lib"
+local git = require "nvim-tree.git"
```
Appears unused.

ghostbuster91 replied:
it is used in the line https://github.com/nvim-tree/nvim-tree.lua/pull/2790/files#diff-b9bfb789bf93265f2e073fa63bcb9783b86a39e42c47b42b1e7193f87a03b489R33

alex-courtis commented on lua/nvim-tree/actions/tree/modifiers/expand-all.lua:
```diff
@@ -18,42 +19,56 @@ local function to_lookup_table(list)
+        if should_expand then
           expansion_count = expansion_count + 1
-          expand(node)
+          populate_node(node)
```
I'm not quite following the logic here, but it looks like we're not actually making many changes e.g.
* `expand` was renamed to `populate_node`
* unnecessary `should_expand` boolean was extracted
* `get_last_group_node` was moved out of `expand`

Please can you refactor this down to the minimum number of changes necessary. Integration with the existing structure will result in simpler code that is more easily reviewable and maintainable.

I appreciate that we've been through many revisions to get to this point, with complexity from earlier commits removed, however we need to revert some structural changes. Perhaps consider how you'd make these changes from scratch, knowing what you have learned over the course of this PR.

ghostbuster91 replied:
> Please can you refactor this down to the minimum number of changes necessary. Integration with the existing structure will result in simpler code that is more easily reviewable and maintainable.

I totally agree with such approach. I think I was able to cut down amount of changes to the bare minimum.

There is one thing that I would like to discuss - changes in the `recursor` function. Before, when the logic was only to expand everything recursively, the condition was to check if `node.open==true` (putting `expansion_count aside`). In my case that is invalid as some nodes have been opened but they shouldn't be traversed further. 

In both cases reusing `should_expand` function results in correct behavior. First of all, this doesn't have to be true in general. Maybe sometimes the logic for `should_recurse` will be different than the logic for `should_expand`. 

Second. So, for each node we call `should_expand` twice, once to check if we should expand and second time to check if we should recurse. Between these calls, if should `should_expand` returns `true`, we call `expand` that changes the state of a given node. This means that users *must not* relay on `node.open` in their `should_expand` implementation as this would result in incorrect behavior. 

I don't have enough experience in developing nvim plugins to judge if this is a problem, but wanted to bring this to your attention. Maybe this should at least be mentioned in the documentation?

alex-courtis replied:
> In both cases reusing should_expand function results in correct behavior. First of all, this doesn't have to be true in general. Maybe sometimes the logic for should_recurse will be different than the logic for should_expand.

I like the consistency, it's more reason-able and will make maintenance easier.

> This means that users _must not_ relay on `node.open` in their `should_expand` implementation as this would result in incorrect behavior.

So long as we document that we are providing a consistent user experience.

Is `node.nodes` correct; can they test that?

alex-courtis reviewed (changes requested):
This is looking good, we'll get there:

- [x] simplify `expand-all.lua` preserving existing structure
- [x] remove unused variables etc. as per luals warnings
- [x] remove any usages of internals like `lib`
- [x] add options to `api.tree.expand_all`, documenting in help and api.lua
- [x] add an example to help, the absolute bare minimum, try to keep to 10 lines. If not, we can add to the wiki.

ghostbuster91 commented:
> ![20240902_121527](https://private-user-images.githubusercontent.com/1505378/363539528-01251e4c-5421-442e-ae05-a3654f5bb234.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUzNDczMjcsIm5iZiI6MTcyNTM0NzAyNywicGF0aCI6Ii8xNTA1Mzc4LzM2MzUzOTUyOC0wMTI1MWU0Yy01NDIxLTQ0MmUtYWUwNS1hMzY1NGY1YmIyMzQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MDNUMDcwMzQ3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZGQ5NDFhNDA4Y2ZhNzI0NWZiMDhkMjU2YTYzYmIzMWJiNTRkNmQ3NzlhOGFjM2QyZjBmMmFmZDk1YTQ1MmFkZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.4NRq89hNHU-juyx-_KgIMHxq_hgKeTya1w-TlhC29PM)
> 
> That parameter is not used and thus can be removed.

Right, I always forget that this is possible in dynamically-typed languages :+1:

ghostbuster91 commented:
> > Any idea how to fix it?
> 
> ![20240902_121129](https://private-user-images.githubusercontent.com/1505378/363539128-ebecf0f8-99bc-4610-ad42-20173fb6ce2b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUzNTcxNTMsIm5iZiI6MTcyNTM1Njg1MywicGF0aCI6Ii8xNTA1Mzc4LzM2MzUzOTEyOC1lYmVjZjBmOC05OWJjLTQ2MTAtYWQ0Mi0yMDE3M2ZiNmNlMmIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MDNUMDk0NzMzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWZiZDZhYzUzODk0NDk1MjhjZWU0ODEyNGVjMWIxYmRmNWRmZjZhYzdiNDM2NTAxOGY0YmQ0MTJkMjVmMmYzYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.49qXxlboRdEZeLnMnQWZK4dZZw6GL1saP8IRaQqobMA)
> 
> status is not a parameter for `function Explorer:expand(node)`, looks like it's unused.

It must have changed once I rebased on latest master. Fixed :+1:

ghostbuster91 commented:
With all these changes my custom should expand function looks as follow:
```lua
local function has_one_child_folder(node)
    return #node.nodes == 1 and node.nodes[1].nodes and
        vim.loop.fs_access(node.nodes[1].absolute_path, "R") or false
end

local function expand_until_non_single(_, node, populate_node)
    populate_node(node)
    if node.nodes == nil or not node.parent.open then
        return false
    end
    return has_one_child_folder(node.parent)
end
```

Thank you for helping me getting this PR across the finish line. I truly appreciate your time and attitude :bowing_man:

alex-courtis commented:
> > status is not a parameter for `function Explorer:expand(node)`, looks like it's unused.
> 
> It must have changed once I rebased on latest master. Fixed 👍

Sorry about that; we are doing some significant refactoring of explorer/reload etc.

alex-courtis commented on lua/nvim-tree/actions/tree/modifiers/expand-all.lua:
```diff
@@ -47,13 +59,14 @@ local function gen_iterator()
       :hidden()
       :applier(function(node)
-        if should_expand(expansion_count, node) then
+        if should_expand(expansion_count, node, populate_node) then
```
![20240907_150214](https://github.com/user-attachments/assets/b6a2c443-7c05-41d0-8506-ee56911c1df7)

This appears unused. Can we remove it?

alex-courtis reviewed (changes requested):
This is looking incredible - the sample gives the user exactly what they need without overwhelming them.

I hope you don't mind: in the interest of speed I've pushed a couple of doc / example tweaks.

- [ ] Issue: Sibling directories are not expanded:

```sh
mkdir -p foo1/bar/baz/foo/bar/baz
mkdir -p foo2/bar/baz/foo/bar/baz
```

`E` at root only expands one level deep:
![20240907_152632](https://github.com/user-attachments/assets/518b2a87-5b43-4d72-a8c8-ec4804921f70)

`Z` at root expands `foo1` correctly but does not expand `foo1`:
![20240907_152726](https://github.com/user-attachments/assets/6b128ec2-c662-4a9c-9903-f57b5c28b322)

ghostbuster91 commented:
@alex-courtis just fyi: I didn't forget about this. Just trying to finish my new apartment, so I can finally move in. I will try to finish this soon.

alex-courtis commented:
Closing as this PR is stale. Please reopen if you wish to continue this work.
